File: easynav/src/easynav_behaviors/easynav_patrolling_behavior/src/easynav_patrolling_behavior/PatrollingNode.cpp
// Copyright 2025 Intelligent Robotics Lab
//
// This file is part of the project Easy Navigation (EasyNav in short)
// licensed under the GNU General Public License v3.0.
// See <http://www.gnu.org/licenses/> for details.
//
// Easy Navigation program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.


#include "rclcpp/rclcpp.hpp"
#include "rclcpp/macros.hpp"

#include "geometry_msgs/msg/pose_stamped.hpp"

#include "easynav_system/GoalManagerClient.hpp"
#include "easynav_patrolling_behavior/PatrollingNode.hpp"

namespace easynav
{

using namespace std::chrono_literals;

PatrollingNode::PatrollingNode(const rclcpp::NodeOptions & options)
: Node("patrolling_node", options)
{
  timer_ = create_timer(
    100ms,
    std::bind(&PatrollingNode::cycle, this));
}

void
PatrollingNode::initialize()
{
  std::vector<std::string> waypoints;
  declare_parameter("waypoints", waypoints);
  get_parameter("waypoints", waypoints);

  declare_parameter<std::string>("frame_id", "map");
  get_parameter("frame_id", frame_id_);

  goals_.header.frame_id = frame_id_;

  for (const auto & wp : waypoints) {
    std::vector<double> wp_coord;
    declare_parameter(wp, wp_coord);
    get_parameter(wp, wp_coord);

    if (wp_coord.size() != 3) {
      RCLCPP_ERROR(get_logger(), "Coordinates for wp [%s] have wrong size %zu",
        wp.c_str(), wp_coord.size());
      continue;
    }

    geometry_msgs::msg::PoseStamped wp_pose;
    wp_pose.header.frame_id = frame_id_;
    wp_pose.pose.position.x = wp_coord[0];
    wp_pose.pose.position.y = wp_coord[1];
    wp_pose.pose.orientation = orientationAroundZAxis(wp_coord[2]);

    goals_.goals.push_back(wp_pose);
  }
}

void
PatrollingNode::cycle()
{
  switch (state_) {
    case PatrolState::IDLE:
      if (!initialized_) {
        RCLCPP_INFO(get_logger(), "Initializing patrolling");

        gm_client_ = GoalManagerClient::make_shared(shared_from_this());
        initialize();
        initialized_ = true;
      }

      RCLCPP_INFO(get_logger(), "Goals sent");

      gm_client_->send_goals(goals_);
      state_ = PatrolState::PATROLLING;
      break;
    case PatrolState::PATROLLING:
      {
        auto nav_state = gm_client_->get_state();
        switch(nav_state) {
          case GoalManagerClient::State::NAVIGATION_REJECTED:
          case GoalManagerClient::State::NAVIGATION_FAILED:
          case GoalManagerClient::State::NAVIGATION_CANCELLED:
          case GoalManagerClient::State::ERROR:
            RCLCPP_ERROR(get_logger(), "Navigation finished with error %s",
              gm_client_->get_result().status_message.c_str());
            state_ = PatrolState::ERROR;
            break;
          case GoalManagerClient::State::NAVIGATION_FINISHED:
            RCLCPP_INFO(get_logger(), "Navigation succesfully finished with message %s",
              gm_client_->get_result().status_message.c_str());
            state_ = PatrolState::FINISHED;
            break;
          case GoalManagerClient::State::ACCEPTED_AND_NAVIGATING:
            RCLCPP_INFO(get_logger(), "Goals accepted and navigating");
            break;
          default:
            break;
        }
      }
      break;
    case PatrolState::FINISHED:
      RCLCPP_INFO(get_logger(), "Reset navigation");
      gm_client_->reset();
      state_ = PatrolState::IDLE;
      break;
    case PatrolState::ERROR:
      break;
  }
}

}  // namespace easynav

File: easynav/src/easynav_behaviors/easynav_patrolling_behavior_py/easynav_patrolling_behavior_py/easynav_patrolling_behavior_py/patrolling_node.py
# Copyright 2025 Intelligent Robotics Lab
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from enum import Enum

import math
from typing import List, Tuple, Optional 

import rclpy
try:
    from easynav_goalmanager_py import GoalManagerClient, ClientState  # type: ignore
except Exception:  # pragma: no cover
    GoalManagerClient = None
    ClientState = None

from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from geometry_msgs.msg import PoseStamped, PoseWithCovarianceStamped
from nav_msgs.msg import Goals
from tf_transformations import quaternion_from_euler

from builtin_interfaces.msg import Time

class PatrolState(Enum):
    IDLE = 0
    PATROLLING = 1
    FINISHED = 2
    ERROR = 3

def quat_from_yaw(yaw: float) -> Tuple[float, float, float, float]:
    cy = math.cos(yaw * 0.5)
    sy = math.sin(yaw * 0.5)
    return (0.0, 0.0, sy, cy)

class PatrollingNode(Node):
    def __init__(self):
        super().__init__('patrolling_node')

        wp = ['']

        self.declare_parameter('frame_id', 'map')
        self.declare_parameter('waypoints', wp)

        self.frame_id: str = self.get_parameter('frame_id').get_parameter_value().string_value

        wp_names = self.get_parameter('waypoints').get_parameter_value().string_array_value
        
        self._goals = Goals()
        self._goals.header.frame_id = self.frame_id

        for name in wp_names:
            vals = [0.0, 0.0, 0.0]
            self.declare_parameter(name, vals)
            param = self.get_parameter(name)
            if not param:
                self.get_logger().warn(f'Missing waypoint parameter: {name}')
                continue
            vals = param.get_parameter_value().double_array_value
            if len(vals) != 3:
                self.get_logger().warn(f'Waypoint {name} must have 3 values [x, y, yaw]')
                continue

            new_wp = PoseStamped()
            new_wp.header.frame_id = self.frame_id
            new_wp.pose.position.x = vals[0]
            new_wp.pose.position.y = vals[1]

            q = quaternion_from_euler(0, 0, vals[2])

            new_wp.pose.orientation.x = q[0]
            new_wp.pose.orientation.y = q[1]
            new_wp.pose.orientation.z = q[2]
            new_wp.pose.orientation.w = q[3]

            self._goals.goals.append(new_wp)

        self._gm = GoalManagerClient(node=self)

        self._wp_index = 0
        self._state = PatrolState.IDLE

        self._timer = self.create_timer(0.2, self._cycle)
        self.get_logger().info('PatrollingNode started')


    def _cycle(self):
        match self._state:
            case PatrolState.IDLE:
                nav_state = self._gm.get_state()

                if nav_state != ClientState.ACCEPTED_AND_NAVIGATING:
                    self.get_logger().debug('Goals sent')
                    self._goals.header.stamp = self.get_clock().now().to_msg()
                    self._gm.send_goals(self._goals)
                else:
                    self._state = PatrolState.PATROLLING
            case PatrolState.PATROLLING:
                nav_state = self._gm.get_state()
                if (
                    nav_state == ClientState.NAVIGATION_REJECTED or
                    nav_state == ClientState.NAVIGATION_FAILED or
                    nav_state == ClientState.NAVIGATION_CANCELLED or
                    nav_state == ClientState.ERROR
                ):
                    result = self._gm.get_result()
                    self.get_logger().debug(
                        f'Navigation finished with error {result.status_message}')
                    self._state = PatrolState.ERROR
                elif nav_state == ClientState.NAVIGATION_FINISHED:
                    result = self._gm.get_result()
                    self.get_logger().info(
                        f'Navigation succesfully finished with message {result.status_message}')
                    self._state = PatrolState.FINISHED
                elif nav_state == ClientState.ACCEPTED_AND_NAVIGATING:
                    self.get_logger().info('Goals accepted and navigating')
                else:
                    pass
            case PatrolState.FINISHED:
                self.get_logger().info('Reset navigation')
                
                nav_state = self._gm.get_state()
                if nav_state != ClientState.IDLE:
                    self._gm.reset()
                else:
                    self._state = PatrolState.IDLE
            case PatrolState.ERROR:
                pass


def main(args=None):
    rclpy.init(args=args)
    node = PatrollingNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()
